### PR TITLE
datapath: Add DSR IPv6 implementation

### PR DESCRIPTION
**Pull request by brb**
This PR:
- Adds the IPv6 DSR implementation.
- Moves DSR-related code (both IPv{4,6}) from `bpf_lxc.c` into `lib/nodeport.h`.

Reviewable per commit.

Tested manually.

This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9895)

### CODE REVIEW

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -119,6 +130,175 @@ static __always_inline int nodeport_nat_ipv6_rev(struct __sk_buff *skb,
...
+	if (skb_adjust_room(skb, sizeof(opt), BPF_ADJ_ROOM_NET, 0))
+		return DROP_INVALID;
+
+	if (skb_store_bytes(skb, ETH_HLEN + sizeof(*ip6), &opt, sizeof(opt), 0) < 0)
+		return DROP_INVALID;
+
+	return 0;
+}
+
+static __always_inline int find_dsr_v6(struct __sk_buff *skb, __u8 nexthdr,
```
Comment (brb) [inline bpf/lib/nodeport.h:160]: Maybe it makes sense to extend `ipv6_hdrlen` by adding a callback function which could be used to find the DSR extension option?

**Comment by brb** [comment]
test-me-please

File: bpf/lib/nodeport.h (hunk 1 above)
Comment (borkmann) [reply bpf/lib/nodeport.h:160]: Maybe if the callback is also a __always_inline it could work with clang so that it inlines everything, haven't tried.

**Review by borkmann — APPROVED** [review]

**Pull request merged**